src/flow/src/batching_mode/task.rs

```diff
@@ -0,0 +1,836 @@
+    sql_to_df_plan, AddAutoColumnRewriter, AddFilterRewriter, FindGroupByFinalName,
+};
+use crate::batching_mode::{MIN_REFRESH_DURATION, SLOW_QUERY_THRESHOLD};
+use crate::error::{
+    ConvertColumnSchemaSnafu, DatafusionSnafu, DatatypesSnafu, ExternalSnafu, InvalidRequestSnafu,
+    SubstraitEncodeLogicalPlanSnafu, TableNotFoundMetaSnafu, TableNotFoundSnafu, UnexpectedSnafu,
+};
+use crate::metrics::{
+    METRIC_FLOW_BATCHING_ENGINE_QUERY_TIME, METRIC_FLOW_BATCHING_ENGINE_SLOW_QUERY,
+};
+use crate::Error;
+
+#[derive(Clone)]
+pub struct BatchingTask {
+    pub flow_id: FlowId,
+    pub query: String,
+    plan: LogicalPlan,
+    pub time_window_expr: Option<TimeWindowExpr>,
+    /// in seconds
+    pub expire_after: Option<i64>,
+    sink_table_name: [String; 3],
+    pub source_table_names: HashSet<[String; 3]>,
+    table_meta: TableMetadataManagerRef,
+    pub state: Arc<RwLock<TaskState>>,
+}
+
+impl BatchingTask {
+    #[allow(clippy::too_many_arguments)]
+    pub fn new(
+        flow_id: FlowId,
+        query: &str,
+        plan: LogicalPlan,
+        time_window_expr: Option<TimeWindowExpr>,
+        expire_after: Option<i64>,
+        sink_table_name: [String; 3],
+        source_table_names: Vec<[String; 3]>,
+        query_ctx: QueryContextRef,
+        table_meta: TableMetadataManagerRef,
+        shutdown_rx: oneshot::Receiver<()>,
+    ) -> Self {
+        Self {
+            flow_id,
+            query: query.to_string(),
+            plan,
+            time_window_expr,
+            expire_after,
+            sink_table_name,
+            source_table_names: source_table_names.into_iter().collect(),
+            table_meta,
+            state: Arc::new(RwLock::new(TaskState::new(query_ctx, shutdown_rx))),
+        }
+    }
+
+    /// Test execute, for check syntax or such
+    pub async fn check_execute(
+        &self,
+        engine: &QueryEngineRef,
+        frontend_client: &Arc<FrontendClient>,
+    ) -> Result<Option<(u32, Duration)>, Error> {
+        // use current time to test get a dirty time window, which should be safe
+        let start = SystemTime::now();
+        let ts = Timestamp::new_second(
+            start
+                .duration_since(UNIX_EPOCH)
+                .expect("Time went backwards")
+                .as_secs() as _,
+        );
+        self.state
+            .write()
+            .await
+            .dirty_time_windows
+            .add_lower_bounds(vec![ts].into_iter());
+
+        if !self.is_table_exist(&self.sink_table_name).await? {
+            let create_table = self.gen_create_table_expr(engine.clone()).await?;
+            info!(
+                "Try creating sink table(if not exists) with expr: {:?}",
+                create_table
+            );
+            self.create_table(frontend_client, create_table).await?;
+            info!(
+                "Sink table {}(if not exists) created",
+                self.sink_table_name.join(".")
+            );
+        }
```

Comment:
Let's enhance `information_schema.flows` to include recent execution results (success/error messages).